### ISSUE TITLE
Pattern reachability algorithm fails in the presence of uninhabited types

### ISSUE THREAD

**Issue by glaebhoerl** [rust-lang/rust#12609]
From `check_match.rs`:

```
// Algorithm from http://moscova.inria.fr/~maranget/papers/warn/index.html
//
// Whether a vector `v` of patterns is 'useful' in relation to a set of such
// vectors `m` is defined as there being a set of inputs that will match `v`
// but not any of the sets in `m`.
//
// This is used both for reachability checking (if a pattern isn't useful in
// relation to preceding patterns, it is not reachable) and exhaustiveness
// checking (if a wildcard pattern is useful in relation to a matrix, the
// matrix isn't exhaustive).

// Note: is_useful doesn't work on empty types, as the paper notes.
// So it assumes that v is non-empty.
```

This results in, most prominently (#4499), a `match` on an empty enum with a wildcard pattern being accepted:

```
enum X { }
fn x() -> X { unimplemented!() }
match x() {
    _ => println!("hi")
}
```

while if you add a variant to the enum and a corresponding arm to the match:

```
enum X { X }
fn x() -> X { unimplemented!() }
match x() {
    X => println!("hello"),
    _ => println!("hi")
}
```

rustc rejects the program with `error: unreachable pattern`.

Clearly a trailing wildcard pattern after you've already matched all possible values in preceding arms should always be considered unreachable, and this is true even when the number of potential values, and hence preceding arms, are both zero.

As the comment notes the algorithm we use relies on "no empty types" as an assumption, so this may not be an _easy_ bug, but it's still a bug.

**Comment by pnkfelix** [comment]
triage: P-medium

**Comment by glaebhoerl** [comment]
FWIW, I've come across two other pieces of work on exhaustiveness checking in pattern matching which may be relevant here. Note that I haven't actually *read* them or investigated very deeply; just to note that they exist and perhaps bring them to others' attention.

One is [this blog post by Neel Krishnawami](http://semantic-domain.blogspot.hu/2012/08/pattern-compilation-made-easy.html), which begins: 

> One of the most practically useful features of functional programming languages is pattern matching, and one of the most important quality-of-implementation features for an implementation is the exhaustiveness checking for pattern matches. In this post, I will give the simplest algorithm for coverage checking that I know.

Later on, it contains the sentence:

> The fifth rule handles the void type. It just succeeds, since there are no patterns of empty type! 

which suggests that this algorithm properly takes account of uninhabited (void) types. (Again though, I haven't taken time to understand the algorithm.)

The other is [a paper called "GADTs meet their match"](http://research.microsoft.com/en-us/um/people/simonpj/papers/pattern-matching/gadtpm.pdf). This is less obviously relevant - there is no explicit mention of uninhabited types - but given that one of the chief challenges of pattern matches with GADTs is accurately determining which cases actually impossible due to their types, it seems like the problems and solutions might at least be something similar.

**Comment by Gankro** [comment]
triage: still a thing; would be cool to have more interesting analysees surrounding uninhabited types like:

```rust
let res: Result<u32, Void> = Ok(0); // same memory layout as plain u32, since Err is unreachable
let Ok(num) = res; // Works because Err is unreachable
```

**Comment by brson** [comment]
Triage: nominating for retriage. Seems like a serious issue without updates in a long time. cc @rust-lang/lang @rust-lang/compiler .

**Comment by aturon** [comment]
Lang team discussed, and we don't feel that this is a very serious issue -- in particular, this only comes up when matching against an empty type, which is very rare to begin with, and the lack of exhaustiveness checking there is unlikely to hide any serious bugs. P-low.

**Comment by glaebhoerl** [comment]
Not just matching directly on an empty (uninhabited) type, but also on a composite type involving one. In particular types like `Result<Foo, !>` might become more common if #1216 lands, and right now the exhaustiveness checker doesn't see that the `Err` variant is impossible. ([Another example.](https://www.reddit.com/r/rust/comments/4e9xb5/error_enum_thoughts/d1yc0j0))

(Not disputing the priority, which I have no opinion on; only the characterization.)

**Comment by aturon** [comment]
@glaebhoerl Yes, my summary was a little to brief -- the match against an empty type could occur in the context of another type.

It is a good point that the introduction of `!` could make this situation more common, but I think that until we start hearing reports that the lack of exhaustiveness checking there is causing trouble in practice, we can safely leave this at P-low.

**Comment by canndrew** [comment]
I think this should be fixed before `feature(never_type)` can be considered stable. In particular, if someone has a `Result<T, !>`, they should be able to match on it with `match res { Ok(v) => v }`

**Comment by glaebhoerl** [comment]
Some concrete motivation for that: for user-defined uninhabited `enum`s (`enum Void { }`), if we want the same behavior, we'd have to go through a warning-cycle or something, at least, before making the change. But `Result<T, !>` is not currently a legal type (in stable), so there would be no backwards compatibility concerns either (as long as we change this behavior *before* making it stable, as @canndrew proposes).

**Comment by nikomatsakis** [comment]
@canndrew I added this as a note to the `!` tracking issue

**Comment by canndrew** [comment]
I've had a crack at implementing this here: https://github.com/rust-lang/rust/pull/36476
It seems to work but definitely needs review from someone who actually understands the code I hacked on.

Also, the change I've implemented *allows* people to omit patterns that are unreachable due to uninhabited types but doesn't *force* them to. This is important for backwards-compatibility but also because I think code like this:

```rust
let res: Result<T, E> = ...;
match res {
    Ok(t) => ...,
    Err(e) => ...,
}
```

should be allowed to compile even when `E` (or `T`) is uninhabited. This might be important for someone who's (eg.) generating code with a macro.
